### PR TITLE
chore: add introspection fetcher with retries

### PR DESCRIPTION
**Pull request by spetrunin**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: packages/sdk/src/definition/introspection-fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+const initAxios = (): AxiosInstance => {
+	const instance = axios.create();
+
+	axiosRetry(instance, {
+		retries: 5,
+		retryDelay: axiosRetry.exponentialDelay,
+		retryCondition: (error: AxiosError) => {
+			if (error.response) {
+				return error.response.status >= 400;
```
Comment (StarpTech) [inline packages/sdk/src/definition/introspection-fetcher.ts:23]: I'd retry only on 5xx and network errors. 40xx means the request is not in the proper state.

File: packages/sdk/src/definition/introspection-fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+			}
+			return true;
+		},
+		onRetry: (retryCount, error, requestConfig) => {
+			console.log(
+				`failed to perform request method: ${requestConfig.method} url: ${requestConfig.url}. Retry attempt #${retryCount}`
```
Comment (StarpTech) [inline packages/sdk/src/definition/introspection-fetcher.ts:29]: Can we make it a bit clear to what this retry belongs? E.g. with a custom log message for the introspection part.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/sdk/src/definition/introspection-fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	return axiosInstance;
+};
+
+const initAxios = (): AxiosInstance => {
+	const instance = axios.create();
+
+	axiosRetry(instance, {
+		retries: 5,
+		retryDelay: axiosRetry.exponentialDelay,
+		retryCondition: (error: AxiosError) => {
```
Comment (StarpTech) [inline packages/sdk/src/definition/introspection-fetcher.ts:21]: You can just use the default behaviour or `axiosRetry.isNetworkOrIdempotentRequestError`

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/sdk/src/definition/introspection-fetcher.ts (hunk 3 above)
Comment (spetrunin) [reply packages/sdk/src/definition/introspection-fetcher.ts:21]: According to this:
https://github.com/softonic/axios-retry/blob/master/es/index.mjs#L19

IsIdempotentRequestError - will return false for `post` request

Comment (spetrunin) [reply packages/sdk/src/definition/introspection-fetcher.ts:21]: yep, checked - no retries in case we use that

Comment (StarpTech) [reply packages/sdk/src/definition/introspection-fetcher.ts:21]: Good point.

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**